File: scala/problem5.scala
/*
https://projecteuler.net/problem=5

Smallest multiple
*/

object problem5 {
  private def gcd(a: Long, b: Long): Long = {
    if (b == 0L) a else gcd(b, a % b)
  }

  private def lcm(a: Long, b: Long): Long = {
    a / gcd(a, b) * b
  }

  def baselineAnswer: Long = {
    (1L to 20L).reduce(lcm)
  }

  def optimizedAnswer: Long = {
    val primes = Array(2, 3, 5, 7, 11, 13, 17, 19)
    var value = 1L

    for (p <- primes) {
      var power = p.toLong
      while (power * p <= 20L) {
        power *= p
      }
      value *= power
    }

    value
  }

  case class Result(value: Long, timeNs: Long)

  def answer: Long = {
    optimizedAnswer
  }

  def eulerBaseline(): Result = {
    val start = System.nanoTime()
    val value = baselineAnswer
    val end = System.nanoTime()
    Result(value, end - start)
  }

  def euler(): Result = {
    val start = System.nanoTime()
    val value = optimizedAnswer
    val end = System.nanoTime()
    Result(value, end - start)
  }

  def main(args: Array[String]): Unit = {
    println("Project Euler Problem 5")

    val baseline = eulerBaseline()
    val optimized = euler()

    println(s"Result (baseline): ${baseline.value}")
    println(s"Time (baseline): ${baseline.timeNs} ns")
    println(s"Result (optimized): ${optimized.value}")
    println(s"Time (optimized): ${optimized.timeNs} ns")

    if (optimized.timeNs > 0) {
      val speedup = baseline.timeNs.toDouble / optimized.timeNs.toDouble
      println(s"Speedup: ${speedup}x")
    }
  }
}

File: cpp/problem5.cpp
/*
https://projecteuler.net/problem=5

Smallest multiple

2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?
*/

#include <chrono>
#include <iostream>
using namespace std;

struct result {
    long long value;
    long long time_ns;
};

long long gcd(long long a, long long b) {
    while (b != 0) {
        long long t = b;
        b = a % b;
        a = t;
    }
    return a;
}

long long lcm(long long a, long long b) {
    return (a / gcd(a, b)) * b;
}

long long evaluate_baseline() {
    long long answer = 1;
    for (int i = 2; i <= 20; ++i) {
        answer = lcm(answer, i);
    }
    return answer;
}

long long evaluate_optimized() {
    const int limit = 20;
    long long answer = 1;

    const int primes[] = {2, 3, 5, 7, 11, 13, 17, 19};
    for (int p : primes) {
        long long max_power = p;
        while (max_power * p <= limit) {
            max_power *= p;
        }
        answer *= max_power;
    }

    return answer;
}

result euler() {
    const auto start = std::chrono::steady_clock::now();
    long long value = evaluate_optimized();
    const auto end = std::chrono::steady_clock::now();
    long long elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count();
    return {value, elapsed};
}

result euler_baseline() {
    const auto start = std::chrono::steady_clock::now();
    long long value = evaluate_baseline();
    const auto end = std::chrono::steady_clock::now();
    long long elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count();
    return {value, elapsed};
}

int main() {
    std::cout << "Project Euler Problem 5" << std::endl;

    result baseline = euler_baseline();
    result optimized = euler();

    std::cout << "Result (baseline): " << baseline.value << std::endl;
    std::cout << "Time (baseline): " << baseline.time_ns << " ns" << std::endl;
    std::cout << "Result (optimized): " << optimized.value << std::endl;
    std::cout << "Time (optimized): " << optimized.time_ns << " ns" << std::endl;

    if (optimized.time_ns > 0) {
        double speedup = static_cast<double>(baseline.time_ns) / static_cast<double>(optimized.time_ns);
        std::cout << "Speedup: " << speedup << "x" << std::endl;
    }

    return 0;
}
